### PR TITLE
Fix #4778: Geokrety login UI should be easier

### LINKED ISSUE #4778: Geokrety login UI should be easier
Currently the user enters his login and password and additionally has to select "Retrieve API key".
This could be redesigned to make it look likt e.g. the authorization to Opencaching sites.
Just enter a username and password and select "Authorize" to complete it.

Not yet sure however about how this thing works.

### PR DESCRIPTION
**Pull request by kumy**

### CODE REVIEW

File: main/res/xml/preferences.xml [hunk 1]
```diff
@@ -352,50 +352,28 @@
                 android:title="@string/init_geokrety" >
                 <PreferenceCategory android:title="@string/settings_settings" >
                     <CheckBoxPreference
-                        android:defaultValue="true"
```
Comment (Bananeweizen) [inline main/res/xml/preferences.xml]: You should separate different issues in different PRs, even if one of them contains only one single line. This line is about having GK not active by default and is completely unrelated to the remaining changes about authorization.

File: main/src/cgeo/geocaching/connector/trackable/GeokretyAuthorizationActivity.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+import ch.boye.httpclientandroidlib.HttpResponse;
+
+import android.os.Bundle;
+
+import java.util.regex.Pattern;
+
+public class GeokretyAuthorizationActivity extends TokenAuthorizationActivity {
+
+    private static final Pattern PATTERN_IS_ERROR = Pattern.compile("error ([\\d]+)");
```
Comment (Bananeweizen) [inline main/src/cgeo/geocaching/connector/trackable/GeokretyAuthorizationActivity.java]: duplicated code, same constants defined in another activity

File: main/src/cgeo/geocaching/connector/trackable/TrackableConnector.java [hunk 3]
```diff
@@ -19,6 +19,8 @@
  */
 public interface TrackableConnector {
 
+    public int getPreferenceActivity();
```
Comment (Bananeweizen) [inline main/src/cgeo/geocaching/connector/trackable/TrackableConnector.java]: When extending interfaces, it is a good idea to add javadoc. We have so little javadoc (meaning we didn't do a good job at this in the past), and it is hard to understand for other developers what is meant with "isRegistered", "isActive", ... etc.

File: main/res/values/strings.xml [hunk 4]
```diff
@@ -580,6 +583,7 @@
     <string name="settings_open_geokretymap_website">Open GeoKretyMap.org website</string>
     <string name="settings_settings">Settings</string>
     <string name="settings_information">Information</string>
+    <string name="settings_enhancement">Enhancement</string>
```
Comment (Bananeweizen) [inline main/res/values/strings.xml]: I'm not even sure what this is referring to, but I believe the term might be wrong, as I can't imagine this to appear anywhere in a settings screen. Is this referring to "Extensions", "Advanced settings", or something similar?

File: main/res/values/strings.xml [hunk 5]
```diff
@@ -236,6 +236,7 @@
     <string name="info_stored_image">New image saved to:</string>
     <string name="info_storing_static_maps">Trying to store static maps</string>
     <string name="info_cache_saved">The cache has been stored locally</string>
+    <string name="err_trackable_log_not_anonymous">Logging trackables from c:geo cannot be anonymous. Please connect to the service.</string>
```
Comment (Bananeweizen) [inline main/res/values/strings.xml]: ...cannot be done anonymously.  or ...cannot be done without user name and password.

"Please connect to the service" is easier to understand when rephrased as "Please add your user name and password in the %s settings", with the service name inserted.

File: main/src/cgeo/geocaching/connector/trackable/AbstractTrackableConnector.java [hunk 6]
```diff
@@ -18,6 +18,9 @@
 public abstract class AbstractTrackableConnector implements TrackableConnector {
 
     @Override
+    public abstract int getPreferenceActivity();
```
Comment (Bananeweizen) [inline main/src/cgeo/geocaching/connector/trackable/AbstractTrackableConnector.java]: Overriding an interface method as an abstract method in a class doesn't make sense (and is reported as code inspection issue in IDEA at least). I suggest to make this a concrete method and to remove the all the "return 0;" implementations from the other connectors.

File: main/src/cgeo/geocaching/activity/TokenAuthorizationActivity.java [hunk 7]
```diff
@@ -0,0 +1,319 @@
+package cgeo.geocaching.activity;
```
Comment (Bananeweizen) [inline main/src/cgeo/geocaching/activity/TokenAuthorizationActivity.java]: What exactly is this class needed for? I thought that we can use our OAuthAuthorization activity for all kinds of OAuth authorization, Geokrety included. I must be missing something here.

File: main/res/xml/preferences.xml (hunk 1 above)
Comment (kumy) [inline main/res/xml/preferences.xml]: I'll try to remember ;)

File: main/src/cgeo/geocaching/connector/trackable/GeokretyAuthorizationActivity.java (hunk 2 above)
Comment (kumy) [inline main/src/cgeo/geocaching/connector/trackable/GeokretyAuthorizationActivity.java]: Are you sure?
This pattern is a specific check for the GK responses from server.
Those types of constant shouldn't be "local", as for strings?

File: main/src/cgeo/geocaching/connector/trackable/TrackableConnector.java (hunk 3 above)
Comment (kumy) [inline main/src/cgeo/geocaching/connector/trackable/TrackableConnector.java]: I'll have to learn javadoc too...

File: main/res/values/strings.xml (hunk 4 above)
Comment (kumy) [inline main/res/values/strings.xml]: I vote +1 for "Advanced settings"

File: main/res/values/strings.xml (hunk 5 above)
Comment (kumy) [inline main/res/values/strings.xml]: GK could logged anonymously directly on geokrety.org. But as we use the API, authentication is mandatory. Users used to log anonymously on GK website may be disturbed.

What about "Logging trackables from c:geo cannot be done anonymously. Please fill your credentials in the %s settings"?

File: main/src/cgeo/geocaching/connector/trackable/AbstractTrackableConnector.java (hunk 6 above)
Comment (kumy) [inline main/src/cgeo/geocaching/connector/trackable/AbstractTrackableConnector.java]: I'll check.

File: main/src/cgeo/geocaching/activity/TokenAuthorizationActivity.java (hunk 7 above)
Comment (kumy) [inline main/src/cgeo/geocaching/activity/TokenAuthorizationActivity.java]: As said in the related issue, GK service doesn't provide OAuth authentication.
I've tried to be as generic as possible. So next connector that need similar authentication system could be easily implemented. This code is largely inspired from your OAuth implementation.

File: main/src/cgeo/geocaching/activity/TokenAuthorizationActivity.java [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+            }
+            requestTokenDialog.show();
+            startButton.setEnabled(false);
+            startButton.setOnTouchListener(null);
+            startButton.setOnClickListener(null);
+
+            RxUtils.networkScheduler.createWorker().schedule(new Action0() {
+                @Override
+                public void call() {
+                    requestToken(usernameEditText.getText().toString(), passwordEditText.getText().toString());
```
Comment (samueltardieu) [inline main/src/cgeo/geocaching/activity/TokenAuthorizationActivity.java]: You are not supposed to access `usernameEditText.getText()` from the network thread. Capture the username and password into `final String` variables right before scheduling the action.

File: main/src/cgeo/geocaching/connector/trackable/GeokretyAuthorizationActivity.java (hunk 2 above)
Comment (samueltardieu) [inline main/src/cgeo/geocaching/connector/trackable/GeokretyAuthorizationActivity.java]: I think too they are better as local variables. I think @Bananeweizen was due to an uncertainty about whether this authentication was derived form OAuth or was completely different, and the comment may be moot now that this has cleared up.

Comment (kumy) [inline main/src/cgeo/geocaching/connector/trackable/GeokretyAuthorizationActivity.java]: Thanks. I let it as is.

**Comment by kumy** [comment]
Thanks for review.

File: main/src/cgeo/geocaching/activity/TokenAuthorizationActivity.java [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+            }
+            requestTokenDialog.show();
+            startButton.setEnabled(false);
+            startButton.setOnTouchListener(null);
+            startButton.setOnClickListener(null);
+
+            final String username = usernameEditText.getText().toString();
+            final String password = passwordEditText.getText().toString();
+
+                RxUtils.networkScheduler.createWorker().schedule(new Action0() {
```
Comment (samueltardieu) [inline main/src/cgeo/geocaching/activity/TokenAuthorizationActivity.java]: There seems to be an indentation issue, lines 192 and 194 should be indented the same way.

**Comment by kumy** [comment]
Fixed the indent issue

File: main/src/cgeo/geocaching/activity/TokenAuthorizationActivity.java [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+        startButton.setText(StringUtils.isBlank(getToken()) ? getAuthStart() : getAuthAgain());
+    }
+
+    @Override
+    public void onNewIntent(final Intent intent) {
+        setIntent(intent);
+    }
+
+    @Override
+    public void onResume() {
```
Comment (samueltardieu) [inline main/src/cgeo/geocaching/activity/TokenAuthorizationActivity.java]: This identical override is useless, please remove.

File: main/src/cgeo/geocaching/activity/TokenAuthorizationActivity.java [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+    public static final int AUTHENTICATED = 1;
+    private static final int ERROR_EXT_MSG = 2;
+
+    @NonNull private String host = StringUtils.EMPTY;
+    @NonNull private String pathToken = StringUtils.EMPTY;
+    @NonNull private String urlRegister = StringUtils.EMPTY;
+    @NonNull private String fieldUsername = StringUtils.EMPTY;
+    @NonNull private String fieldPassword = StringUtils.EMPTY;
+
+    private boolean https = false;
```
Comment (samueltardieu) [inline main/src/cgeo/geocaching/activity/TokenAuthorizationActivity.java]: Better use something that can be identified as `boolean`, such as `useHTTPS`. Nevertheless, why single out http/https and pass it as an extra, rather than getting it from the URL? And why decompose the URL into components at all?

Comment (kumy) [inline main/src/cgeo/geocaching/activity/TokenAuthorizationActivity.java]: I'll check it this tonight.
Note: code is inspired from AuthorizationActivity.java
The same remark could be done for https://github.com/cgeo/cgeo/blob/master/main/src/cgeo/geocaching/activity/OAuthAuthorizationActivity.java#L56 ? ;)

File: main/src/cgeo/geocaching/activity/TokenAuthorizationActivity.java (hunk 10 above)
Comment (kumy) [inline main/src/cgeo/geocaching/activity/TokenAuthorizationActivity.java]: I agree.

File: main/src/cgeo/geocaching/activity/TokenAuthorizationActivity.java (hunk 11 above)
Comment (samueltardieu) [inline main/src/cgeo/geocaching/activity/TokenAuthorizationActivity.java]: Yes, the same remark can certainly be done for the code you cite. It should probably be cleaned up as well in the future.

**Comment by kumy** [comment]
Removed the https parameter and define the full url for each case.
Code rebased on current master.
